backends/audiocd/macosx/macosx-audiocd.cpp

```diff
@@ -0,0 +1,295 @@
+	kMaxDriveCount = 256
+};
+
+MacOSXAudioCDManager::DriveList MacOSXAudioCDManager::detectAllDrives() {
+	// Fetch the lists of drives
+	struct statfs driveStats[kMaxDriveCount];
+	int foundDrives = getfsstat(driveStats, sizeof(driveStats), MNT_WAIT);
+	if (foundDrives <= 0)
+		return DriveList();
+
+	DriveList drives;
+	for (int i = 0; i < foundDrives; i++)
+		drives.push_back(Drive(driveStats[i].f_mntonname, driveStats[i].f_mntfromname, driveStats[i].f_fstypename));
+
+	return drives;
+}
+
+bool MacOSXAudioCDManager::play(int track, int numLoops, int startFrame, int duration, bool onlyEmulate) {
+	// Prefer emulation
+	if (DefaultAudioCDManager::play(track, numLoops, startFrame, duration, onlyEmulate))
+		return true;
+
+	// If we're set to only emulate, or have no CD drive, return here
+	if (onlyEmulate || !_trackMap.contains(track))
+		return false;
+
+	if (!numLoops && !startFrame)
+		return false;
+
+	// Now load the AIFF track from the name
+	Common::String fileName = _trackMap[track];
+	Common::SeekableReadStream *stream = StdioStream::makeFromPath(fileName.c_str(), false);
+
+	if (!stream) {
+		warning("Failed to open track '%s'", fileName.c_str());
+		return false;
+	}
+
+	Audio::AudioStream *audioStream = Audio::makeAIFFStream(stream, DisposeAfterUse::YES);
+	if (!audioStream) {
+		warning("Track '%s' is not an AIFF track", fileName.c_str());
+		return false;
+	}
+
+	Audio::SeekableAudioStream *seekStream = dynamic_cast<Audio::SeekableAudioStream *>(audioStream);
+	if (!seekStream) {
+		warning("Track '%s' is not seekable", fileName.c_str());
+		return false;
+	}
+
+	Audio::Timestamp start = Audio::Timestamp(0, startFrame, 75);
+	Audio::Timestamp end = duration ? Audio::Timestamp(0, startFrame + duration, 75) : seekStream->getLength();
+
+	// Fake emulation since we're really playing an AIFF file
+	_emulating = true;
+
+	_mixer->playStream(Audio::Mixer::kMusicSoundType, &_handle,
+	                   Audio::makeLoopingAudioStream(seekStream, start, end, (numLoops < 1) ? numLoops + 1 : numLoops), -1, _cd.volume, _cd.balance);
+	return true;
+}
+
+bool MacOSXAudioCDManager::findTrackNames(const Common::String &drivePath) {
+	Common::FSNode directory(drivePath);
+
+	if (!directory.exists()) {
+		warning("Directory '%s' does not exist", drivePath.c_str());
+		return false;
+	}
+
+	if (!directory.isDirectory()) {
+		warning("'%s' is not a directory", drivePath.c_str());
+		return false;
+	}
+
+	Common::FSList children;
+	if (!directory.getChildren(children, Common::FSNode::kListFilesOnly)) {
+		warning("Failed to find children for '%s'", drivePath.c_str());
+		return false;
+	}
+
+	for (uint32 i = 0; i < children.size(); i++) {
+		if (!children[i].isDirectory()) {
+			Common::String fileName = children[i].getName();
+
+			if (fileName.hasSuffix(".aiff") || fileName.hasSuffix(".cdda")) {
+				uint trackID = 0, j = 0;
+
+				for (; j < fileName.size() && !Common::isDigit(fileName[j]); j++)
```

Comment:
Aha, yes, I see! This may be one of those very rare cases where it makes sense to actually comment what the code is doing, so dumb people like me don’t gloss over the one-character difference. But it’s up to you, and I could change my mind tomorrow when I’m not drunk-tired and trying to be “helpful” reviewing unfamiliar stuff :)